Allow masking threads out at runtime (#4615)
numba/numba: docs/source/developer/threading_implementation.rst

=========================================
Notes on Numba's threading implementation
=========================================

The execution of the work presented by the Numba ``parallel`` targets is
undertaken by the Numba threading layer. Practically, the "threading layer"
is a Numba built-in library that can perform the required concurrent execution.
At the time of writing there are three threading layers available, each
implemented via a different lower level native threading library. More
information on the threading layers and appropriate selection of a threading
layer for a given application/system can be found in the
:ref:`threading layer documentation <numba-threading-layer>`.

The pertinent information to note for the following sections is that the
function in the threading library that performs the parallel execution is the
``parallel_for`` function. The job of this function is to both orchestrate and
execute the parallel tasks.

Thread masking
--------------

> In order to simplify the design, it was decided that Numba should never launch

Author: So are you saying this restriction of never launching new threads after the initial launch could be removed in the future, assuming someone wants to implement what would be required to make it work?

> In order to simplify the design, it was decided that Numba should never launch

Reviewer: Correct, I think that this restriction could be removed. With thanks to the current design I also think it'd be quite easy to implement (famous last words pending)...

1. Apply restrictions so that only the TBB and OpenMP backends can be used with this behaviour (or rewrite `workqueue` to permit it). I think this reduces to:
   a. Remove restrictions in `set_num_threads()`:
      * Add a "backend" check to be able to check which backend is running
      * If the backend is TBB or OpenMP remove the masking restriction.

   b. I don't think there's anything more to be done. `get_num_threads` should work as-is, the rest of the implementation should be ok.
2. Do something with `_launch_threads()` and the initial threadpool size, this might be "launching" a default size to begin with, such that `NUMBA_NUM_THREADS` simply prescribes the default.
3. Probably worth implementing a `NUMBA_THREAD_LIMIT` or something so as to prevent dynamically creating more threads than cores by default, this can obviously be altered by the user.

new threads beyond the threads that are launched initially with
``_launch_threads`` when the first parallel execution is run. Consequently,
the programmatic setting of the number of threads can only be done by setting

> the number of threads to a number less than the total number that have already

Reviewer: How about:
```
it's possible to programmatically set the number of threads but only to less than or equal to the total number that have already been launched.
```
?

do no work. For example, on a 16 core machine, if the user were to call
``set_num_threads(4)``, numba would always have 16 threads present, but 12 of
them would sit idle for parallel computations. A further call to
``set_num_threads(16)`` would cause those same threads to do work in later
computations.

:ref:`Thread masking <numba-threading-layer-thread-masking>` was added to make
it possible for a user to programmatically alter the number of threads
performing work in the threading layer. Thread masking proved challenging to
implement as it required the development of a programming model that is suitable
for users, easy to reason about, and could be implemented safely, with
consistent behavior across the various threading layers.

Programming model
~~~~~~~~~~~~~~~~~

The programming model chosen is similar to that found in OpenMP. The reasons
for this choice were that it is familiar to a lot of users, restricted in
scope and also simple. The number of threads in use is specified by calling
``set_num_threads`` and the number of threads in use can be queried by calling
``get_num_threads``.These two functions are synonymous with their OpenMP

> counterparts (with the above restriction that the mask must be <= the number

Reviewer: ```suggestion
counterparts (with the above restriction that the mask must be less than or equal to the number
```

of launched threads). The execution semantics are also similar to OpenmP in
that once a parallel region is launched altering the thread mask, it has no
impact on the currently executing region but will have an impact on parallel

> regions executed subsequently.

Reviewer: I'm not sure that this makes sense. The launch of a parallel region has no impact on the thread mask, the original text was correct:
```
The execution semantic is also similar to OpenMP in that once a
parallel region is launched, altering the thread mask has no impact on the
currently executing region, but will have an impact on parallel regions executed
subsequently.
```
?

i.e. once in a `parallel for` loop, setting the thread mask won't do anything to the executing loop, but a subsequent (logical or nested) `parallel for` loop will adhere to the new mask.

Author: Oh my bad. I think the missing comma before "altering" confused me.

Reviewer: Ah, probably my bad, I should have put the comma in in the first place! Anyway, looks good now, thanks. Sorry for the confusion!

  different mask with which to call nested functions, if it's not set explicitly
  then the inherited mask from the "main thread" is used
* threading layers which have dynamic scheduling with threads potentially
  joining and leaving the active pool during a ``parallel_for`` execution are
  successfully accommodated
* any "main thread" thread mask is entirely decoupled from the in-flux nature
  of the thread masks of the threads in the active thread pool

Python threads independently invoking parallel functions
********************************************************

The threading layer launch sequence is heavily guarded to ensure that the
launch is both thread and process safe and run once per process. In a system
with numerous Python ``threading`` module threads all using Numba, the first
thread through the launch sequence will get its thread mask set appropriately,
but no further threads can run the launch sequence. This means that other
threads will need their initial thread mask set some other way. This is
achieved when ``get_num_threads`` is called and no thread mask is present, in
this case the thread mask will be set to the default. In the implementation,

> "no thread mask is present" is represented by the value -1 and the "default

Reviewer: ```suggestion
"no thread mask is present" is represented by the value ``-1`` and the "default
```

> thread mask" (unset) is represented by the value 0. The implementation also

Reviewer: ```suggestion
thread mask" (unset) is represented by the value ``0``. The implementation also
```

immediately calls ``set_num_threads(NUMBA_NUM_THREADS)`` after doing this, so

> if either -1 or 0 is encountered as a result from ``get_num_threads()`` it

Reviewer: ```suggestion
if either ``-1`` or ``0`` is encountered as a result from ``get_num_threads()`` it
```

indicates a bug in the above processes.

OS ``fork()`` calls
*******************

The use of TLS was also in part driven by the Linux (the most popular
platform for Numba use by far) having a ``fork(2, 3P)`` call that will do TLS
propagation into child processes, see ``clone(2)``\ 's ``CLONE_SETTLS``.

Thread ID
*********

A private ``get_thread_id()`` function was added to each threading backend,
which returns a unique ID for each thread. This can be accessed from Python by
``numba.npyufunc.parallel._get_thread_id()`` (it can also be used inside of an

> njitted function). The thread ID function is useful for testing that the

Reviewer: Perhaps `... used from within a JIT compiled function`, `njitted` tends not to be used.

thread masking behavior is correct, but it should not be used outside of the
tests. For example, one can call ``set_num_threads(4)`` and then collect all
unique ``_get_thread_id()``\ 's in a parallel region to verify that only 4
threads are run.

Caveats
~~~~~~~

> Some caveats to be aware of when testing this:

Reviewer: Given this is a new section, to what does `this` apply?

- The TBB backend may choose to schedule fewer than the given mask number of
  threads. Thus a test such as the one described above may return fewer than 4
  unique threads.

- The workqueue backend is not threadsafe, so attempts to do nested

>   parallelism with it may result in deadlocks or other undefined behavior.

Reviewer: Is this not now caught and a `SIGABRT` raised?

Author: My understanding was that the SIGABRT would catch nested threading, but would it also catch other attempts to multi-thread workqueue, which aren't nested?

Reviewer: It would catch nesting only, it wouldn't prevent the deadlock/segfault that happens if the workqueue backend is accessed concurrently. What you have added in the latest patches looks good with respect to this.

- Certain backends may reuse the main thread for computation, but this

>   behavior shouldn't be relied on (for instance, for exceptions propagating).

Reviewer: ```suggestion
  behavior shouldn't be relied upon (for instance, if propagating exceptions).
```

Use in Code Generation
~~~~~~~~~~~~~~~~~~~~~~

The general pattern for using ``get_num_threads`` in code generation is

.. code:: python

   import llvmlite.llvmpy.core as lc

   get_num_threads = builder.module.get_or_insert_function(
       lc.Type.function(lc.Type.int(types.intp.bitwidth), []),
       name="get_num_threads")

   num_threads = builder.call(get_num_threads, [])

   with cgutils.if_unlikely(builder, builder.icmp_signed('<=', num_threads,
                                                 num_threads.type(0))):
       cgutils.printf(builder, "num_threads: %d\n", num_threads)
       context.call_conv.return_user_exc(builder, RuntimeError,
                                                 ("Invalid number of threads. "
                                                  "This likely indicates a bug in Numba.",))

   # Pass num_threads through to the appropriate backend function

See the code in ``numba/npyufunc/parfor.py``. Here ``builder.module`` is the

> thread pool backend library, e.g., ``tbbpool``.

Reviewer: `builder.module` is the currently active LLVM module being built, the code above is emitting instructions into the module to stage a call to `get_num_threads` and check the result.

Author: I see. I misunderstood how `get_or_insert_function` worked.

Reviewer: I'm wondering if some tutorial material would be useful for `builder` things? It is a bit unusual!

12 of them would sit idle for parallel computations. A further call to
``set_num_threads(16)`` would cause those same threads to do work in later
computations.

:ref:`Thread masking <numba-threading-layer-thread-masking>` was added to make
it possible for a user to programmatically alter the number of threads
performing work in the threading layer. Thread masking proved challenging to
implement as it required the development of a programming model that is suitable
for users, easy to reason about, and could be implemented safely, with
consistent behavior across the various threading layers.

Programming model
~~~~~~~~~~~~~~~~~

The programming model chosen is similar to that found in OpenMP. The reasons
for this choice were that it is familiar to a lot of users, restricted in
scope and also simple. The number of threads in use is specified by calling
``set_num_threads`` and the number of threads in use can be queried by calling
``get_num_threads``.These two functions are synonymous with their OpenMP
counterparts (with the above restriction that the mask must be less than or
equal to the number of launched threads). The execution semantics are also

> similar to OpenmP in that once a parallel region is launched, altering the

Reviewer 2: ```suggestion
similar to OpenMP in that once a parallel region is launched, altering the
```

thread through the launch sequence will get its thread mask set appropriately,
but no further threads can run the launch sequence. This means that other
threads will need their initial thread mask set some other way. This is
achieved when ``get_num_threads`` is called and no thread mask is present, in
this case the thread mask will be set to the default. In the implementation,
"no thread mask is present" is represented by the value ``-1`` and the "default
thread mask" (unset) is represented by the value ``0``. The implementation also
immediately calls ``set_num_threads(NUMBA_NUM_THREADS)`` after doing this, so
if either ``-1`` or ``0`` is encountered as a result from ``get_num_threads()`` it
indicates a bug in the above processes.

OS ``fork()`` calls
*******************

The use of TLS was also in part driven by the Linux (the most popular
platform for Numba use by far) having a ``fork(2, 3P)`` call that will do TLS
propagation into child processes, see ``clone(2)``\ 's ``CLONE_SETTLS``.

Thread ID
*********

A private ``get_thread_id()`` function was added to each threading backend,
which returns a unique ID for each thread. This can be accessed from Python by

> ``numba.npyufunc.parallel._get_thread_id()`` (it can also be used inside of

Reviewer 2: ```suggestion
``numba.npyufunc.parallel._get_thread_id()`` (it can also be used inside a
```

JIT compiled function). The thread ID function is useful for testing that the
thread masking behavior is correct, but it should not be used outside of the
tests. For example, one can call ``set_num_threads(4)`` and then collect all

> unique ``_get_thread_id()``\ 's in a parallel region to verify that only 4

Reviewer 2: ```suggestion
unique ``_get_thread_id()``\ s in a parallel region to verify that only 4
```

threads will need their initial thread mask set some other way. This is
achieved when ``get_num_threads`` is called and no thread mask is present, in
this case the thread mask will be set to the default. In the implementation,
"no thread mask is present" is represented by the value ``-1`` and the "default
thread mask" (unset) is represented by the value ``0``. The implementation also
immediately calls ``set_num_threads(NUMBA_NUM_THREADS)`` after doing this, so

> if either ``-1`` or ``0`` is encountered as a result from ``get_num_threads()`` it

Reviewer: long line, but we can catch things like this in the actual review.
